model/address.rb

```diff
@@ -5,8 +5,44 @@
 class Address < Sequel::Model
   one_to_many :assigned_vm_addresses, key: :address_id, class: :AssignedVmAddress
   one_to_many :assigned_host_addresses, key: :address_id, class: :AssignedHostAddress
+  many_to_one :vm_host, key: :routed_to_host_id
 
   plugin ResourceMethods
+
+  def validate
+    super
+
+    if new? && cidr.is_a?(NetAddr::IPv4Net) && cidr.len > 256
+      errors.add(:cidr, "too large (contains more than 256 addresses)")
+    end
+  end
+
+  def before_destroy
+    DB[:ipv4_address].where(cidr:).delete
+    super
+  end
+
+  def after_create
+    super
+    populate_ipv4_addresses
+  end
+
+  def populate_ipv4_addresses
+    # Do nothing for ipv6 addresses, since VM addresses are chosen randomly from the /64.
+    # Ignore /32 IPv4 addresses, since those would be used by the host itself and not for
+    # VMs running on the host.
+    if cidr.is_a?(NetAddr::IPv4Net) && id != routed_to_host_id
+      addresses = Array.new(cidr.len) { [cidr.nth(it), cidr.to_s] }
+
+      if vm_host.provider_name == "leaseweb"
+        # Do not use first or last addresses for leaseweb
+        addresses.shift
+        addresses.pop
+      end
+
+      DB[:ipv4_address].import([:ip, :cidr], addresses)
+    end
```

Comment:
It's totally personal (feel free to ignore) but I prefer early return instead of wrapping whole function with a block. Less indentation


```suggestion
    return unless cidr.is_a?(NetAddr::IPv4Net) && id != routed_to_host_id
    
    addresses = Array.new(cidr.len) { [cidr.nth(it), cidr.to_s] }

    if vm_host.provider_name == "leaseweb"
      # Do not use first or last addresses for leaseweb
      addresses.shift
      addresses.pop
    end

    DB[:ipv4_address].import([:ip, :cidr], addresses)
```